add cache utility

jackstar12 opened the pull request:
This pr introduces a simple redis-like cache utility.

### Possible use cases:

- temporary caching of exchange rates, especially usefull with something like https://github.com/lnbits/lnbits/pull/1789
- caching node peer information or payment history in cln at https://github.com/lnbits/lnbits/pull/1702
- improving api key validation which currently queries the entire wallet, including balance, see: https://github.com/lnbits/lnbits/blob/bc55d52ea25475a524059fccbc5a35039f8be533/lnbits/decorators.py#L52-L55

callebtc commented:
cACK

jackstar12 commented:
something like this could be used alternatively: https://pypi.org/project/expiring-dict/

arcbtc commented:
LGTM!
Caching exchange info would be very useful

motorina0 commented on lnbits/cache.py:
```diff
@@ -0,0 +1,67 @@
+                self.pop(key, prefix)
+        return None
+
+    def set(self, key: str, value: Any, expiry: float = 10, prefix: str = ""):
```
`prefix` should go after `key` (for consistency with `get`)

motorina0 commented on lnbits/cache.py:
```diff
@@ -0,0 +1,67 @@
+    def __init__(self):
+        self._values: dict[Any, Cached] = {}
+
+    def get(self, key: str, prefix: str = "") -> Optional[Any]:
```
why no `defaultValue` (like `pop()`)

callebtc replied:
The cache is an unordered key-value store... what would you expect with an argument-less `pop()`?

motorina0 commented on lnbits/cache.py:
```diff
@@ -0,0 +1,67 @@
+
+    def pop(self, key: str, prefix: str = "", default=None) -> Optional[Any]:
+        cached = self._values.pop(_add_prefix(key, prefix), None)
+        if cached:
```
shouldn't `cached.expiry > time()` also be checked here?

motorina0 commented on lnbits/cache.py:
```diff
@@ -0,0 +1,67 @@
+            return cached.value
+        return default
+
+    async def save_result(self, coro, key: str, expiry: float = 10, prefix: str = ""):
```
- order of params should be consistent with the other methods
- what is the advantage of this functionality? 
   - `coro` can be called at the line before `save_results` and the results passed to `save_results`
   - it only saves one line of code

motorina0 replied:
- if `coro` has params then a wrapper haws to be created
- I would remove this function, keep the interface as simple as possible

jackstar12 replied:
> * it only saves one line of code

```python
rate = cache.get("ccy-usd")
if not rate:
    rate = await fetch_rate("usd")
    cache.set("ccy-usd", rate)
```
vs
```python
rate = await cache.save_result(lambda: fetch_rate("usd") key="ccy-usd")
```

callebtc commented:
LGTM except for the prefix, either I don't get it or it looks redundant to me.

motorina0 commented on lnbits/cache.py:
```diff
@@ -0,0 +1,67 @@
+            try:
+                await asyncio.sleep(interval)
+                ts = time()
+                self._values = {k: v for k, v in self._values.items() if v.expiry > ts}
```
- `invalidate_forever` runs as part of an `asyncio` task
- `invalidate_forever` has a `for` loop

What happens if an element is removed from `self._values.items()` (via `pop`) during the iteration?
Will it throw `"RuntimeError: dictionary changed size during iteration`?

jackstar12 replied:
thats not possible because asyncio is single-threaded

motorina0 commented on tests/core/test_cache.py:
```diff
@@ -0,0 +1,61 @@
+@pytest.mark.asyncio
+async def test_cache_get_set(cache):
+    cache.set(key, value)
+    assert cache.get(key) == value
```
assert that calling `cache.get()` twice will still return `value`

jackstar12 commented:
> Cache should not loop over all dictionary entries at every time step and not create a new dictionary to return. Instead use the common pattern for implementing cache:
> 
> * set value
> * kick of async task that waits
> * remove value

I adjusted the task to run every 10s instead of 1 and removed the recreation of the dict - which should be fine. Looping over dict with 100,000 entries takes 4ms on my machine. I think asyncio thanks us aswell for not scheduling thousands of callbacks

jackstar12 commented:
In regards to `save_result`. I could remove it but I think its a pretty common pattern. I already use it in https://github.com/lnbits/lnbits/pull/1702 and it would fit really well into the currency rates and wallet info caching aswell

callebtc reviewed (approved):
LGTM
